ipc_service: Extend API with nocopy functions

carlocaione opened the pull request:
This PR follows the discussion at #43924. The idea is that we want to extend the IPC service API to expose several new functions that can be used by backends with nocopy capabilities.

In this PR we also adapt the static vrings backend and the related sample.

carlocaione committed:
ipc_service: Extend API with nocopy functions

Add several new functions to the IPC service API:
  - ipc_service_get_tx_buffer()
  - ipc_service_drop_tx_buffer()
  - ipc_service_send_nocopy()
  - ipc_service_release_rx_buffer()
  - ipc_service_hold_rx_buffer()

This set of function is used to support backends with nocopy capability.

Signed-off-by: Carlo Caione <ccaione@baylibre.com>

hubertmis commented on include/ipc/ipc_service.h:
```diff
@@ -170,6 +170,80 @@ int ipc_service_register_endpoint(const struct device *instance,
+ *  @retval 0 on success.
+ *  @retval other errno codes depending on the implementation of the backend.
+ */
+int ipc_service_get_buffer(struct ipc_ept *ept, void **data, uint32_t *len, int wait);
```
nit: docs state that `wait` type is boolean
```suggestion
int ipc_service_get_buffer(struct ipc_ept *ept, void **data, uint32_t *len, bool wait);
```

nordic-krch replied:
would it make sense to have `k_timeout_t` here?

carlocaione replied:
ok

hubertmis commented on include/ipc/ipc_service.h:
```diff
@@ -170,6 +170,80 @@ int ipc_service_register_endpoint(const struct device *instance,
+ *  @retval -EIO when no backend is registered or send hook is missing from
+ *	    backend.
+ *  @retval -EINVAL when instance or endpoint is invalid.
+ *  @retval -EBADMSG when the message is invalid.
```
What message?

carlocaione replied:
the data message. This is currently returned for example in the static vrings backend when you try to send an empty message.

hubertmis replied:
Could you explain it in more details in the comment? I didn't think that any message could be considered invalid while reading this documentation.

carlocaione replied:
why not? It really depends on the backend. Maybe it doesn't support zero-length messages like in our case or maybe the backend expects some kind of data formatting (some protocol?) and it wants to inform the user when something is wrong with that.

hubertmis replied:
I understand it now. But I needed more details. I'm asking you to add these details to the doc in the header file to help readers understand what could be considered "invalid message"

carlocaione replied:
I'll do.

hubertmis commented on include/ipc/ipc_service.h:
```diff
@@ -170,6 +170,80 @@ int ipc_service_register_endpoint(const struct device *instance,
+ *  using @ref ipc_service_get_buffer and filling the TX buffer with the data.
+ *
+ *  After the @ref ipc_service_send_nocopy function is issued the TX buffer is
+ *  no more owned by the sending task and must not be touched anymore unless
+ *  the function fails and returns an error.
```
It looks a function to release a TX buffer is needed. If a user calls `ipc_service_send_nocopy()` which returns one of the error responses, user may choose to discard the buffer.

arnopo replied:
Right, 
A related issue as just be entered in OpenAMP:
https://github.com/OpenAMP/open-amp/issues/361

pdunaj replied:
I would not add a function but assume the backend will always take ownership regardless if send is successful or not.

The only reason to keep it separated would be if the user was responsible for send retry. But we need retry only if buffer is not available which is not the case here -we already grabbed a buffer earlier. I don't see a point in complicating the API.

carlocaione replied:
We need a separate `send()` function for this because the backend must be able to differentiate the nocopy case vs the regular case. See for example OpenAMP functions `rpmsg_send_offchannel_nocopy()` vs `rpmsg_send_offchannel()`.

hubertmis commented on include/ipc/ipc_service.h:
```diff
@@ -170,6 +170,80 @@ int ipc_service_register_endpoint(const struct device *instance,
+ */
+int ipc_service_send_nocopy(struct ipc_ept *ept, const void *data, size_t len);
+
+/** @brief Release the RX buffer for future reuse.
```
I would split RX nocopy feature to other commit than TX nocopy (for clear separation).
For RX nocopy we need two functions like
* `rx_buffer_keep()`
* `rx_buffer_throw_away()`

To use zero-copy feature, user needs to call `rx_buffer_keep()` in `received()` callback implementation. When user stops using this buffer `rx_buffer_throw_away()` should be called.

If user does not call `rx_buffer_keep()` in `received()` callback implementation, typical copying operation is assumed and the buffer is automatically released after `received()` callback processing.

nordic-krch replied:
maybe `claim` and `release` would be better.

carlocaione replied:
ok, I'll try to extend the RX path API.

nordic-krch commented on include/ipc/ipc_service.h:
```diff
@@ -170,6 +170,80 @@ int ipc_service_register_endpoint(const struct device *instance,
+ *
+ *  @param ept Registered endpoint by @ref ipc_service_register_endpoint.
+ *  @param data Pointer to the empty TX buffer.
+ *  @param len Pointer to store the TX buffer size.
```
Shouldn't it be an input where you provide requested length? Or [in,out] where on output granted length (equal or bigger than requested) is stored.

nordic-krch replied:
but #43544 is using circular buffer so you need to know how much space to allocate because there is a variable length allocation.

emob-nordic replied:
I think we need only one parameter (`reqested`) one (for me its okay if its named `len`), and this function should return `-ENOBUFS` or `-ENOMEM` if requested buffer can not be claimed withing `k_timeout_t timeout`. I think there is no point to add `returned` size, because user needs specific size, if cannot be provided error must be returned.

nordic-krch commented on include/ipc/ipc_service.h:
```diff
@@ -170,6 +170,80 @@ int ipc_service_register_endpoint(const struct device *instance,
+ *	    backend.
+ *  @retval -EINVAL when instance or endpoint is invalid.
+ *  @retval -ENOTSUP when this is not supported by backend.
+ *  @retval -ENOBUFS when there are no TX buffers available.
```
what about `-EALREADY` when buffer cannot be granted because one is already claimed. That may be the case for implementations which are using circular buffer like #43544.

carlocaione replied:
uhm, in theory you can keep requesting buffers until there are available buffers, so you are not forced to claim one single buffer at time. When there are no more available buffers you get `-ENOBUFS` so I don't think that having `-EALREADY` is useful or needed.

nordic-krch replied:
in case of #43544 it will not support multiple `get_buffer` calls, `-ENOBUFS` would indicate that circular buffer is full and `-EALREADY` would indicate that `get_buffer` was called and it is not yet sent. In case of backend that supports multiple `get_buffers` `-EALREADY` would never be returned.

carlocaione replied:
ok makes sense

nordic-krch commented on include/ipc/ipc_service.h:
```diff
@@ -170,6 +170,80 @@ int ipc_service_register_endpoint(const struct device *instance,
+ */
+int ipc_service_send_nocopy(struct ipc_ept *ept, const void *data, size_t len);
+
+/** @brief Release the RX buffer for future reuse.
```
i don't understand this feature. Is this the buffer that was a parameter to the callback? Does it mean that buffer is not released when callback returns and must be released by this function and it can be done even after callback returns?

carlocaione replied:
correct.

carlocaione commented on include/ipc/ipc_service.h:
```diff
@@ -170,6 +170,80 @@ int ipc_service_register_endpoint(const struct device *instance,
+ *  ipc_service_send_nocopy function to perform data no-copy-send mechanism.
+ *
+ *  @param ept Registered endpoint by @ref ipc_service_register_endpoint.
+ *  @param data Pointer to the empty TX buffer.
```
ACK

pdunaj commented on include/ipc/ipc_service_backend.h:
```diff
@@ -71,6 +71,63 @@ struct ipc_service_backend {
+	 *  @param token Backend-specific token.
+	 *  @param data Pointer to the empty TX buffer.
+	 *  @param len Pointer to store the TX buffer size.
+	 *  @param wait To wait or not for the buffer to become available.
```
this cannot be a simple boolean - let's use timeout for that

carlocaione replied:
ok

pdunaj commented on include/ipc/ipc_service.h:
```diff
@@ -170,6 +170,80 @@ int ipc_service_register_endpoint(const struct device *instance,
+ *  @param ept Registered endpoint by @ref ipc_service_register_endpoint.
+ *  @param data Pointer to the empty TX buffer.
+ *  @param len Pointer to store the TX buffer size.
+ *  @param wait Boolean, wait or not for buffer to become available
```
let's use timeout here

pdunaj commented on include/ipc/ipc_service.h:
```diff
@@ -170,6 +170,80 @@ int ipc_service_register_endpoint(const struct device *instance,
+ *  has been obtained by using @ref ipc_service_get_buffer.
+ *
+ *  The application has to take the responsibility for getting the TX buffer
+ *  using @ref ipc_service_get_buffer and filling the TX buffer with the data.
```
let's also mention about order of sending data. The simplest approach (I recommend it) is to allow only one message in progress. That is, after you get the buffer you MUST send a message before you are allowed to get another buffer.
Moreover you must do both accesses from the same context.

If we don't limit it this way the complexity of the mechanism will from (for now good reason). Keeping track of all started messages and handling synchronization between threads will make this code heavy (for no good reason).

carlocaione replied:
I'll add a comment.

carlocaione committed:
ipc_service: Use dir indication in parameters

And add some more details in documentation.

Signed-off-by: Carlo Caione <ccaione@baylibre.com>

hubertmis commented on include/ipc/ipc_service.h:
```diff
@@ -66,22 +122,29 @@ extern "C" {
+	 *  This callback is called when new data is received.
+	 *
+	 *  @note The data buffer is to be considered released and available
+	 *        again only when this callback returs.
```
```suggestion
	 *        again only when this callback returns.
```

carlocaione replied:
oops

hubertmis commented on include/ipc/ipc_service.h:
```diff
@@ -157,21 +222,154 @@ int ipc_service_register_endpoint(const struct device *instance,
+
+/** @brief Drop and release a TX buffer
+ *
+ *  When @ref ipc_service_send_nocopy returns an error, it is possible to use
```
Is it ok to use this function to drop a buffer without calling `ipc_serivce_send_nocopy` earlier?

I mean if a sequence like this is valid:

```
r = ipc_service_get_tx_buffer(...);

if (r == 0 && something_happend_causing_the_user_is_not_interested_in_sending_anymore()) {
  r = ipc_service_drop_tx_buffer(...);
}
```

From the description I understand it is not valid.

carlocaione replied:
I think the API should allow that. I'll clarify that use case.

hubertmis commented on include/ipc/ipc_service.h:
```diff
@@ -157,21 +222,154 @@ int ipc_service_register_endpoint(const struct device *instance,
+ */
+int ipc_service_release_rx_buffer(struct ipc_ept *ept, const void *data);
+
+/** @brief Holds the RX buffer for usage outside the receive callback.
```
It would be easier to follow this API if `hole_rx_buffer()` was described before `release_rx_buffer()`.

I'm missing an information here if implementation is supposed to count number of `hold_rx_buffer()` calls and release the buffer only when matching number of `release_rx_buffer()` is called, or is it like a binary status.

Let me show an example:
```
void received_callback(...) {
  ipc_service_hold_rx_buffer(...);
  ...
}

void other_processing(void) {
  ipc_service_hold_rx_buffer(...);

  ipc_service_release_rx_buffer(...); // Is buffer released here?
  ipc_service_release_rx_buffer(...); // Or here?
}
```

carlocaione replied:
so
1) `hold` and `release` are acting on the `data` parameter:
```
ipc_service_hold_rx_buffer(A);
ipc_service_hold_rx_buffer(B);
...
ipc_service_release_rx_buffer(B); // This is releasing buffer B
ipc_service_release_rx_buffer(A); // This is releasing buffer A
```
2) You cannot release the same buffer multiple times, otherwise you get `-EALREADY`:
```
ipc_service_hold_rx_buffer(A);
ipc_service_hold_rx_buffer(A); # here you get -EALREADY
...
ipc_service_release_rx_buffer(A);
ipc_service_release_rx_buffer(A); # here you get -EALREADY
```

I will add a new error case for when you try to `release` a buffer that was not `hold`-ed.

pdunaj replied:
> I'm missing an information here if implementation is supposed to count number of hold_rx_buffer() calls and release the buffer only when matching number of release_rx_buffer() is called, or is it like a binary status.

I mentioned it earlier. I would add restriction that there is only one buffer being held and only by one owner. Otherwise this will become complex. Note that practically there will be always one context using the buffer.

hubertmis replied:
I meant calling `hold_rx_buffer()` multiple times for the same buffer, so restricting number of buffers would not answer my question.

But `-EALREADY` returned by `hold_rx_buffer()` clarifies it.

nordic-krch commented on subsys/ipc/ipc_service/ipc_service.c:
```diff
@@ -81,3 +81,119 @@ int ipc_service_send(struct ipc_ept *ept, const void *data, size_t len)
+
+	backend = ept->instance->api;
+
+	if (!backend || !backend->get_tx_buffer) {
```
when backend has no `get_tx_buffer` pointer shouldn't it return `-ENOTSUP`?

carlocaione replied:
If you are implementing a backend and you do not want to implement `get_rx_buffer` for example you can add the hook and return `-ENOTSUP`. The code in `ipc_service.c` is a failsafe to catch the cases when the user didn't implement the hook to avoid an application crash.

carlocaione committed:
ipc_service: static_vrings: Add NOCOPY capability

Add the newly introduced NOCOPY capability to the static vrings backend.

Signed-off-by: Carlo Caione <ccaione@baylibre.com>

carlocaione committed:
sample: static_vrings: Use NOCOPY capability.

At least for one of the instances, use the NOCOPY feature of the
backend.

Signed-off-by: Carlo Caione <ccaione@baylibre.com>

carlocaione commented:
@nordic-krch @pdunaj @hubertmis @arnopo I added some commits to this PR to also adapt the static vrings backend to the new API and changed the current sample so that we use the NOCOPY on one of the tested instances.

nordic-krch commented on samples/subsys/ipc/ipc_service/static_vrings/src/main.c:
```diff
@@ -179,16 +179,23 @@ K_THREAD_DEFINE(ipc0B_thread_id, STACKSIZE, ipc0B_entry, NULL, NULL, NULL, PRIOR
 static void ipc1_ept_recv(const void *data, size_t len, void *priv)
 {
-	ipc1_received_data = *((uint8_t *) data);
+	ipc_service_hold_rx_buffer(&ipc1_ept, (void *) data);
```
handle error code. In future there might be backend that does not support it.

carlocaione replied:
ack

nordic-krch commented on samples/subsys/ipc/ipc_service/static_vrings/src/main.c:
```diff
@@ -231,14 +237,27 @@ static void ipc1_entry(void *dummy0, void *dummy1, void *dummy2)
-		message = ipc1_received_data;
+
+		message = *((uint8_t *) recv_data);
+		ipc_service_release_rx_buffer(&ipc1_ept, recv_data);
```
handle error code.

carlocaione replied:
ack

nordic-krch commented on subsys/ipc/ipc_service/backends/ipc_rpmsg_static_vrings.c:
```diff
@@ -448,10 +469,86 @@ static int open(const struct device *instance)
+	(*data) = payload;
+
+	if (*len > max_size) {
+		*len = max_size;
```
isn't this a problem? We allocated too small buffer and there is no option to return it  (dropping tx buffer is not supported).
Is there a way to get `max_size` prior to calling `rpmsg_get_tx_payload_buffer` and return error if it is less than requested value?

carlocaione replied:
no. But this will be not an issue when OpenAMP will support the buffer drop (see https://github.com/OpenAMP/open-amp/issues/361)

nordic-krch commented on subsys/ipc/ipc_service/backends/ipc_rpmsg_static_vrings.c:
```diff
@@ -448,10 +469,86 @@ static int open(const struct device *instance)
+
+	/*
+	 * We still want to return the payload buffer in case of -ENOMEM
+	 * because the user would want to drop it or just use it anyway with
```
`... want to drop...` but how? tx drop not supported.

carlocaione replied:
not supported yet. This will be added in OpenAMP soon-ish.

hubertmis commented on subsys/ipc/ipc_service/backends/ipc_rpmsg_static_vrings.c:
```diff
@@ -448,10 +469,86 @@ static int open(const struct device *instance)
+
+	/* OpenAMP only supports a binary wait / no-wait */
+	if (!K_TIMEOUT_EQ(wait, K_FOREVER) && !K_TIMEOUT_EQ(wait, K_NO_WAIT)) {
+		return -EINVAL;
```
`-EINVAL` should be returned only for invalid parameter values, and according to API description any timeout value is valid.
I think this function should implement some kind of retry mechanism if timeout has other value then FOREVER of NO_WAIT.

carlocaione replied:
I really wanted to avoid that TBH because I think that will open a can of worms. What about adding a new error instead for unsupported timeout?

hubertmis replied:
I think `-ENOTSUP` or something similar would be fine. Retry could be implemented later

hubertmis commented on include/ipc/ipc_service.h:
```diff
@@ -203,6 +239,145 @@ int ipc_service_register_endpoint(const struct device *instance,
+ *  The size parameter can be used to request a buffer with a certain size. If
+ *  the size can be accomodated the function returns no errors and the size
+ *  value is unchanged. If the requested size is too big, the function returns
+ *  -ENOMEM and the size parameter must contain the maximum allowed size.
```
I assume the buffer is allocated only when this function returns 0. When it returns other error codes, the buffer is not allocated and the caller is not responsible to drop it. I think this function doc should clearly describe if my assumption is correct.

carlocaione replied:
So we have two possibilities here:
1. If the size cannot be accomodate the function returns `-ENOMEM` and no buffer is allocated so the caller is not responsible to drop it. This could be a problem with OpenAMP because AFAIK there is no way no determine the size without allocation the buffer (***)
2. The function returns `-ENOMEM` but the buffer is allocated anyway and the maximum size is still returned to the caller using the `size` parameter. This is the solution implemented in the static vrings backend. https://github.com/zephyrproject-rtos/zephyr/blob/89ef492572edfccb6efcabd28ba7e2f6a8297aac/subsys/ipc/ipc_service/backends/ipc_rpmsg_static_vrings.c#L497-L507

*** @arnopo is there an OpenAMP API to get the `len` returned by `rpmsg_get_tx_payload_buffer()` without actually allocating the buffer?

hubertmis replied:
I think this API does not have to be fully aligned with OpenAMP. If OpenAMP has this limitation, implementation here could call `rpmsg_free_tx_buffer()` (coming soon) if we decide to not allocate an buffer in `-NOMEM` scenario.

carlocaione replied:
I'm still unclear which way we want to go. Whether we want to still allocate and return the buffer or not.

I think an even better solution could be make the `len` parameter a `desirable` buffer size. If the backend could accomodate that ok, in all the other cases (the backend is able to allocate either a smaller or a bigger buffer) the function should return `0` (because the buffer was allocated) and the allocated size in `*len`. The user can then decide if that's fine (and use the buffer with the returned size) or if he needs to drop it.

@hubertmis @nordic-krch what do you think?

hubertmis replied:
As a user of this API I would like it to allocate the size I'm asking for. If implementation cannot give me what I'm asking, I would expect an error like `-ENOMEM`, possibly accompanied with an information what smaller buffer size could have been allocated.
As API user most probably I'm not interested in this smaller buffer size, because the data I need to send would not fit. Because of this I would prefer to receive `-ENOMEM` and not get any buffers allocated.
If for any reason I could use smaller buffer, I can use the accompanying information to request again.

carlocaione replied:
ok fair enough.

carlocaione commented:
@hubertmis @nordic-krch some changes:
- added a new `ipc_service_get_tx_buffer_size()` function
- `ipc_service_get_tx_buffer()` now fails with `-ENOMEM` when you request a buffer too big / small and the buffer is not allocated. You can still pass `0` as `size` parameter to allocate a buffer as big as possible.

hubertmis commented on include/ipc/ipc_service.h:
```diff
@@ -203,6 +239,169 @@ int ipc_service_register_endpoint(const struct device *instance,
  */
 int ipc_service_send(struct ipc_ept *ept, const void *data, size_t len);
 
+/** @brief Get the TX buffer size
```
Description of this function is ambiguous. How about something like
"Get the maximal size of a buffer which can be obtained now by ipc_service_get_tx_buffer()"

hubertmis commented on subsys/ipc/ipc_service/backends/ipc_rpmsg_static_vrings.c:
```diff
@@ -448,10 +469,107 @@ static int open(const struct device *instance)
+		}
+
+		/* Not the requested size */
+		if (*size != buf_size) {
+			*size = buf_size;
+			return -ENOMEM;
+		}
```
I'm not sure if this implementation works correctly. Let's say I'm developing a module sending 12 bytes of data between two cores. I use an endpoint registered in ipc_service for this purpose. I'm calling `get_tx_buffer(..., 12, ...);` sequentially to allocate buffers for my message. But every time I call this function I receive `-ENOMEM` even though virtio buffers could easily hold my 12 bytes long message. When I try again, I receive `-ENOMEM`, and again, and again.

I think it should be more like `if (*size > buf_size) { return -ENOMEM; } else { /* allocated buffer is larger than requested, but it is fine */ ... return 0; }`

carlocaione replied:
ok, so you want to return `-ENOMEM` only when the requested size is bigger than the buffer size.

hubertmis replied:
I'm thinking if there is any consequence in releasing buffer procedures.

In tx buffer releasing scenario:
```
uint32_t len = 16;
void *data;
r = ipc_service_get_tx_buffer(ept, &data, &len, NO_WAIT); // Allocated buffer is 256 bytes long, function returns 0
...
r = ipc_service_drop_tx_buffer(ept, data);
```
There is no buffer size parameter passed to `drop_tx_buffer()` so probably there is no problem.

In sending tx buffer scenario:
```
uint32_t len = 16;
void *data;
r = ipc_service_get_tx_buffer(ept, &data, &len, NO_WAIT); // Allocated buffer is 256 bytes long, function returns 0
...
r = ipc_service_send_nocopy(ept, data, 16);
```
Allocated buffer is 256 bytes long, but `send_nocopy()` function uses only 16 bytes out of this buffer. I think it is fine to assume that `send_nocopy()` will release whole 256 bytes here anyway.

What do you think?

hubertmis replied:
Then yes, I would prefer `-ENOMEM` only when available size is smaller than requested. If allocated buffer is bigger `0` seems better.

hubertmis commented on samples/subsys/ipc/ipc_service/static_vrings/src/main.c:
```diff
@@ -231,14 +242,32 @@ static void ipc1_entry(void *dummy0, void *dummy1, void *dummy2)
 
 	while (message < 100) {
-		ret = ipc_service_send(&ipc1_ept, &message, sizeof(message));
+		uint32_t len = 0;
```
If this line was changed to `len = sizeof(message);` you would capture the problem I've described.

hubertmis commented on samples/subsys/ipc/ipc_service/static_vrings/remote/src/main.c:
```diff
@@ -233,14 +244,32 @@ static void ipc1_entry(void *dummy0, void *dummy1, void *dummy2)
+			break;
+		}
+
+		*((unsigned char *) data) = message;
```
If `len = 0`, you should verify `len` value after calling `ipc_service_get_tx_buffer()` and before filling the buffer.

hubertmis commented on samples/subsys/ipc/ipc_service/static_vrings/src/main.c:
```diff
@@ -231,14 +242,32 @@ static void ipc1_entry(void *dummy0, void *dummy1, void *dummy2)
 
 	while (message < 100) {
-		ret = ipc_service_send(&ipc1_ept, &message, sizeof(message));
+		uint32_t len = sizeof(message);
```
I think it would be good to use `uint32_t len = 0;` in one communication direction, to verify both requesting buffer of given size, or maximal buffer size.

carlocaione replied:
Could I just leave this as is? If I force push I'll lose the two ACKs and I have to beg people for those again.

hubertmis replied:
Sure, the sample can be improved later

hubertmis reviewed (dismissed):
Looks good!

emob-nordic commented on include/ipc/ipc_service.h:
```diff
@@ -157,21 +223,190 @@ int ipc_service_register_endpoint(const struct device *instance,
+ *  @retval -EIO when no backend is registered or send hook is missing from
+ *		 backend.
+ *  @retval -EINVAL when instance or endpoint is invalid.
+ *  @retval -ENOTSUP when the operation or the timeout is not supported by backend.
```
timeout is probably leftover. Could yo remove it?

carlocaione replied:
damn

arnopo commented on include/ipc/ipc_service.h:
```diff
@@ -66,22 +87,29 @@ extern "C" {
+	 *  This callback is called when new data is received.
+	 *
+	 *  @note The data buffer is to be considered released and available
+	 *        again only when this callback returs.
```
s/returs/returns/

carlocaione replied:
yeah, that was fixed by a following commit in the same PR.

arnopo commented on subsys/ipc/ipc_service/ipc_service.c:
```diff
@@ -81,3 +81,138 @@ int ipc_service_send(struct ipc_ept *ept, const void *data, size_t len)
+		return -EIO;
+	}
+
+	if (!backend->send_nocopy) {
```
test backend->get_tx_buffer here and adapt the message :
		LOG_ERR("No copy service not available");

arnopo replied:
same comments can be applied on other "no copy" APIs

carlocaione replied:
This check is actually doing something slightly different. It's just making sure that we cannot have a `get_tx_buffer()` function without a corrispettive `send_nocopy()` function (and viceversa). A bit later we also check that we cannot have an `hold_rx_buffer()` function without a `release_rx_buffer()` call.

arnopo replied:
Yes this is my point. What about following re-split?

```
if (!backend ) {
	/* The backend reference is not valid */
	return -EINVAL;
}
if (!backend->send_nocopy || !backend->get_tx_buffer) {
		LOG_ERR("No-copy service not available");
		return -EIO;
}
```
And same for `hold_rx_buffer()` & `release_rx_buffer()`

arnopo commented on samples/subsys/ipc/ipc_service/static_vrings/remote/src/main.c:
```diff
@@ -181,16 +181,28 @@ K_THREAD_DEFINE(ipc0B_thread_id, STACKSIZE, ipc0B_entry, NULL, NULL, NULL, PRIOR
+		printk("ipc_service_hold_rx_buffer failed with ret %d\n", ret);
+	}
+
+	recv_data = (void *) data;
```
Here if two call back are received before you release the RX buffer the recv_data will be overwritten.
I suggest to add a comment explaining that the example is designed to support only a synchronous
"request-answer" mechanism. for asynchronous a chain list should be implemented.

carlocaione replied:
ok

arnopo commented on samples/subsys/ipc/ipc_service/static_vrings/remote/src/main.c:
```diff
@@ -233,14 +244,32 @@ static void ipc1_entry(void *dummy0, void *dummy1, void *dummy2)
 		k_sem_take(&ipc1_data_sem, K_FOREVER);
-		message = ipc1_received_data;
+
+		message = *((uint8_t *) recv_data);
```
Any reason to make a copy? why not print it, copy in TX, and then release the buffer?
As it is a synchronous "request-answer" should not lead to a lock

carlocaione replied:
definitely makes sense.

arnopo reviewed (changes requested):
My apologize I'm quite late on the review...
First of all, congratulation for this very good work !
Few minor remarks + a proposal to save a copy

carlocaione commented on include/ipc/ipc_service.h:
```diff
@@ -203,6 +239,174 @@ int ipc_service_register_endpoint(const struct device *instance,
+ *  ipc_service_send_nocopy function to perform data no-copy-send mechanism.
+ *
+ *  The size parameter can be used to request a buffer with a certain size:
+ *  - if the size can be accomodated the function returns no errors and the
```
ok

carlocaione commented on subsys/ipc/ipc_service/ipc_service.c:
```diff
@@ -81,3 +81,138 @@ int ipc_service_send(struct ipc_ept *ept, const void *data, size_t len)
+
+	backend = ept->instance->api;
+
+	if (!backend || !backend->get_tx_buffer) {
```
why do you want to get rid of this? It's basically following the same pattern for the other functions in the same file: we are returning here `-EIO` when the function is not implemented by the backend (that can still returns `-ENOTSUP` though).

arnopo replied:
My comments was not clear,  please refer to https://github.com/zephyrproject-rtos/zephyr/pull/44033#discussion_r840512905
